File: 1.two-sum.js
/*
 * [1] Two Sum
 *
 * https://leetcode.com/problems/two-sum/description/
 *
 * algorithms
 * Easy (39.35%)
 * Total Accepted:    1.2M
 * Total Submissions: 3.1M
 * Testcase Example:  '[2,7,11,15]\n9'
 *
 * Given an array of integers, return indices of the two numbers such that they
 * add up to a specific target.
 * 
 * You may assume that each input would have exactly one solution, and you may
 * not use the same element twice.
 * 
 * Example:
 * 
 * 
 * Given nums = [2, 7, 11, 15], target = 9,
 * 
 * Because nums[0] + nums[1] = 2 + 7 = 9,
 * return [0, 1].
 * 
 * 
 * 
 * 
 */
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
// var twoSum = function(nums, target) {
//     if (nums.length < 2) return [];
//     if (nums.length === 2 && nums[0] + nums[1] === target) return [0, 1];
//     var result = [];
//     for (var i = 0; i < nums.length; i++) {
        
//     }
// };

const twoSum = function (nums, target) {
    const comp = {};
    for (let i = 0; i < nums.length; i++) {
        if (comp[nums[i]] >= 0) {
            return [comp[nums[i]], i]
        }
        comp[target - nums[i]] = i
    }
};

File: 1.two-sum.go
/*
 * @lc app=leetcode id=1 lang=golang
 *
 * [1] Two Sum
 * 思路：
 * 利用一个 hashmap
 * key: array value
 * value: array index
 * 遍历数组
 * 如果 hashmap[target - value] 存在
 * 返回当前 索引 和 hashmap[target - value]
 */

package leetcode

// @lc code=start
func twoSum(nums []int, target int) []int {
	targetMap := make(map[int]int, len(nums))

	for i, value := range nums {
		if j, ok := targetMap[target-value]; ok {
			return []int{j, i}
		}

		targetMap[value] = i
	}

	return nil
}

// @lc code=end
